### PR TITLE
Improve Salt Bundle promote pipeline

### PR DESCRIPTION
**Pull request by ycedres**
This PR fixes the part of https://github.com/SUSE/spacewalk/issues/21338 regarding improvements for the Salt Bundle promote pipeline. 

**Previous behaviour**
Run one `osc copypac` for every dependency package and something similar for subproject configurations. Every new dependency or subproject has to be manually included.
**From now on**
The list of packages and subprojects is retrieved from OBS and the copies are generated dinamycally. If the package is a link to another package it will not be copied. A subproject could reference another subproject, so if the link is copied, the destination project would reference a subproject of the source project.

### CODE REVIEW

**Comment by ycedres** [comment]
I believe we could do something similar for the other pipeline, however I've noticed there that not all packages are copied but just a subset and I'm not sure why.

**Review by meaksh — CHANGES_REQUESTED** [review]
Thanks for the PR. I just left some comments.

BTW, are you planning to push the changes for the classic Salt package pipeline on this same PR or in a differnt one?

File: promote_salt_bundle_packages.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    project_handler = Project(client)
+    project_handler.set_config(prj, config=config)
+
+def has_link(client, project, package) -> bool:
+    pkg_handler = Package(client)
+    pkg_files = pkg_handler.get_files(project,package)
+    linkinfo = etree.fromstring(etree.tostring(pkg_files).decode('utf-8')).find("linkinfo")
+    return linkinfo is not None
+
+import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (ycedres) [reply promote_salt_bundle_packages.py:55]: For sure 😅

File: promote_salt_bundle_packages.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+def copy_packages(client, src, dst, subproject=None):
+    pkg_handler = Package(client)
+    if subproject is not None:
+        src = src + ":" + subproject
+        dst = dst + ":" + subproject
+
+    packages_list = pkg_handler.get_list(project=src)
+    for package in packages_list.iter():
+        package_name = package.attrib.get("name")
+        if package_name is not None and not has_link(client, src, package_name):
```
[the remark this answers is not in the document]
Comment (ycedres) [reply promote_salt_bundle_packages.py]: I added your suggested comments and also a header comment.

**Comment by ycedres** [comment]
> Thanks for the PR. I just left some comments.
> 
> BTW, are you planning to push the changes for the classic Salt package pipeline on this same PR or in a differnt one?

I'd like to push the changes for the classic Salt in a different PR, as I'd like to try this change first and see how it goes.

**Review by meaksh — APPROVED** [review]
LGTM - keep in mind this would require we add `osctiny` as installed package in our Jenkins workers (sumadocker hosts).

We would need to add this new package to these instances via our infrastructure repo, but I just noticed it is already there: https://gitlab.suse.de/galaxy/infrastructure/-/blob/master/srv/salt/jenkins-node/base.sls#L46

So, in principle, we are good to go.

**Comment by meaksh** [comment]
As mentioned in Slack, we just realized this PR is missing showing the "diff" for the project configs.

A new PR will come shortly to including this changes too.

**Pull request merged**